hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/BlobDeleteHandler.java

```diff
@@ -0,0 +1,180 @@
+
+  private boolean nonRecursiveDeleteDirectoryFailed = false;
+
+  private final TracingContext tracingContext;
+
+  private final AtomicInteger deleteCount = new AtomicInteger(0);
+
+
+  public BlobDeleteHandler(final Path path,
+      final boolean recursive,
+      final AbfsBlobClient abfsBlobClient,
+      final TracingContext tracingContext) {
+    super(path, abfsBlobClient, abfsBlobClient.getAbfsConfiguration()
+        .getBlobDeleteDirConsumptionParallelism(), tracingContext);
+    this.path = path;
+    this.recursive = recursive;
+    this.tracingContext = tracingContext;
+  }
+
+  private boolean deleteInternal(final Path path)
+      throws AzureBlobFileSystemException {
+    abfsClient.deleteBlobPath(path, null, tracingContext);
+    deleteCount.incrementAndGet();
+    return true;
+  }
+
+  /**
+   * Orchestrate the delete operation.
+   *
+   * @return true if the delete operation is successful.
+   * @throws IOException if deletion fails due to server error or path doesn't exist.
+   */
+  public boolean execute() throws IOException {
+    /*
+     * ABFS is not aware if it's a file or directory. So, we need to list the
+     * path and delete the listed objects. The listing returns the children of
+     * the path and not the path itself.
+     */
+    listRecursiveAndTakeAction();
+    if (nonRecursiveDeleteDirectoryFailed) {
+      throw new IOException("Non-recursive delete of non-empty directory");
+    }
+    tracingContext.setOperatedBlobCount(deleteCount.get() + 1);
+    /*
+     * If path is actually deleted.
+     */
+    boolean deleted;
+    try {
+      /*
+       * Delete the required path.
+       * Directory needs to be safely delete the path, as the path can be implicit.
+       */
+      deleted = recursive ? safeDelete(path) : deleteInternal(path);
+    } finally {
+      tracingContext.setOperatedBlobCount(null);
+    }
+    if (deleteCount.get() == 0) {
+      /*
+       * DeleteCount can be zero only if the path does not exist.
+       */
+      throw new AbfsRestOperationException(HTTP_NOT_FOUND,
+          PATH_NOT_FOUND.getErrorCode(), PATH_NOT_FOUND.getErrorMessage(),
+          new PathIOException(path.toString(), "Path not found"));
+    }
+
+    /*
+     * Ensure that parent directory of the deleted path is marked as a folder. This
+     * is required if the parent is an implicit directory (path with no marker blob),
+     * and the given path is the only child of the parent, the parent would become
+     * non-existing.
+     */
+    if (deleted) {
+      ensurePathParentExist();
+    }
+    return deleted;
+  }
+
+  private void ensurePathParentExist()
+      throws AzureBlobFileSystemException {
+    AzureBlobFileSystemStore.Permissions permissions
+        = new AzureBlobFileSystemStore.Permissions(false,
+        FsPermission.getDefault(), FsPermission.getUMask(
+        abfsClient.getAbfsConfiguration().getRawConfiguration()));
+    if (!path.isRoot() && !path.getParent().isRoot()) {
```

Comment:
path.getParent() null check should also be there

Reply:
Thats what path.isRoot would do, in case path.isRoot() is true, it means that path.getParent would be null.